### PR TITLE
tests: Ensure all core files are required

### PR DESCRIPTION
**Pull request by bfontaine**
This is a work in progress; I’m still not able to locally get the same coverage as Travis. cc @UniqMartin

### CODE REVIEW

File: Library/Homebrew/test/testing_env.rb [hunk 1]
```diff
@@ -20,6 +20,25 @@
...
 end
 
+if ENV["HOMEBREW_TESTS_COVERAGE"]
+  # Ensure all Ruby files are `require`d for an accurate coverage measurement
+  Dir[File.expand_path("../../**/*.rb", __FILE__)].each do |f|
+    next if [
+      # scripts
+      "build.rb", "postinstall.rb", "test.rb",
+      # mocked files
+      "config.rb", "global.rb",
```
Comment (mikemcquaid) [inline Library/Homebrew/test/testing_env.rb]: I think it would still be good to work out a way of getting these all included and/or put them into designation directories.

File: Library/Homebrew/test/testing_env.rb [hunk 2]
```diff
@@ -20,6 +20,25 @@
...
+    ].any? do |core_file|
+      f.end_with?("Homebrew/#{core_file}")
```
Comment (mikemcquaid) [inline Library/Homebrew/test/testing_env.rb]: When wouldn't they end with that?

Comment (bfontaine) [inline Library/Homebrew/test/testing_env.rb]: `cmd/config.rb` and  `cmd/test.rb` conflict with `config.rb` and `test.rb`.

Comment (mikemcquaid) [inline Library/Homebrew/test/testing_env.rb]: Ah, ok, I see, it's a single check. Found that a bit confusing; could perhaps tweak the readability but just excluding based on the full (or relative) paths?

File: Library/Homebrew/test/testing_env.rb (hunk 1 above)
Comment (bfontaine) [inline Library/Homebrew/test/testing_env.rb]: What do you mean by “designation directories”? We could e.g. mark them as non-coverable with `# :nocov:`.

File: Library/Homebrew/test/testing_env.rb [hunk 3]
```diff
@@ -20,6 +20,25 @@
...
+    end
+
+    # ignore test files
+    next if f.include?("/test/") || f.include?("/vendor/bundle/") || f.end_with?("cmd/test.rb")
```
Comment (UniqMartin) [inline Library/Homebrew/test/testing_env.rb]: The last one in this list should have been `cmd/tests.rb`? Why do we need an extra check for `/vendor/bundle/`? Isn't that a subdirectory of `test` and already covered by `/test/`?

File: Library/Homebrew/test/testing_env.rb (hunk 2 above)
Comment (bfontaine) [inline Library/Homebrew/test/testing_env.rb]: I refactored a little bit; let me know if it’s better.

**Comment by bfontaine** [comment]
I killed the build; no need to run the CI on this for now.

File: Library/Homebrew/test/testing_env.rb (hunk 3 above)
Comment (bfontaine) [inline Library/Homebrew/test/testing_env.rb]: Thanks, I removed both checks. `cmd/test.rb` was erroring at the beginning but now it works.

**Comment by UniqMartin** [comment]
May I suggest a slightly modified version? It makes the path checks much easier.

```ruby
if ENV["HOMEBREW_TESTS_COVERAGE"]
  # Ensure all Ruby files are `require`d for an accurate coverage measurement
  library_base = File.expand_path("../..", __FILE__)
  Dir["#{library_base}/**/*.rb"].each do |f|
    f_rel = f[(library_base.length + 1)..-1] # path relative to `library_base`
    next if %w[build.rb postinstall.rb test.rb].include?(f_rel) # scripts
    next if %w[config.rb global.rb].include?(f_rel) # mocked files
    next if f_rel == "cmd/tests.rb" || f_rel.start_with?("test/") # test files

    require f
  end
end
```

**Comment by bfontaine** [comment]
@UniqMartin What is the secret of the low coverage? :smile:

**Comment by mikemcquaid** [comment]
:+1: to @UniqMartin's version.

**Comment by UniqMartin** [comment]
> What is the secret of the low coverage? :smile:

The issue is with Coveralls and the way we include it in our `.simplecov`, which overwrites the default HTML output formatter of SimpleCov, but then outputs nothing because it isn't running within CI (when we do local testing). The only reason we still get an output to `index.html` in this scenario is because while running the integration tests we do not enable Coveralls and thus get an output (and the coverage) from when the last integration test finishes.

The results cache is only updated with the final results after the last unit test finishes, but with Coveralls enabled `index.html` isn't regenerated. Because this cache is reused on a second run, it now contains the results from the previous unit test run and the coverage is much higher.

This is what fixed it for me locally and should continue to work on Travis:

```diff
diff --git i/Library/Homebrew/test/.simplecov w/Library/Homebrew/test/.simplecov
index 34d22184..e23a1256 100644
--- i/Library/Homebrew/test/.simplecov
+++ w/Library/Homebrew/test/.simplecov
@@ -24,7 +24,9 @@ if name = ENV["HOMEBREW_INTEGRATION_TEST"]
   end
 end

-if RUBY_VERSION.split(".").first.to_i >= 2 && !ENV["HOMEBREW_INTEGRATION_TEST"]
+# Don't use Coveralls outside of CI, as it will override SimpleCov's default
+# formatter causing the `index.html` not to be written once all tests finish.
+if RUBY_VERSION.split(".").first.to_i >= 2 && !ENV["HOMEBREW_INTEGRATION_TEST"] && ENV["CI"]
   require "coveralls"
   Coveralls.wear!
 end
```

**Comment by UniqMartin** [comment]
The issue became obvious once I tried running `brew tests --coverage --verbose TEST=test_ARGV.rb` with a clean cache and got no output at all.

**Comment by bfontaine** [comment]
Awesome, thank you @UniqMartin! I looked at the results cache but never thought about Coveralls breaking stuff.

File: Library/Homebrew/test/testing_env.rb [hunk 4]
```diff
@@ -20,6 +20,25 @@
...
+    ]
+
+    next if excluded_core_files.any? { |file| f.end_with?("Homebrew/#{file}") }
+
+    # ignore test files
+    next if f.include?("/test/")
+
+    require f
+  end
+end
```
Comment (xu-cheng) [inline Library/Homebrew/test/testing_env.rb]: We don't need this right?

Comment (bfontaine) [inline Library/Homebrew/test/testing_env.rb]: We need it otherwise some files are missing from the coverage report.

Comment (xu-cheng) [inline Library/Homebrew/test/testing_env.rb]: I would like to avoid this. Becuase this feels hack to me and likely to cause trouble: e.g. There may be conflicts, the order of files during require matters. 

If it doesn't missed out any coverage. I would be OK to keep the original behavior. If a file is missing from report, we know the coverage for that file is zero right?

File: Library/Homebrew/test/testing_env.rb (hunk 1 above)
Comment (mikemcquaid) [inline Library/Homebrew/test/testing_env.rb]: Sorry, I mean having these files in directories that indicates they are e.g. scripts/configuration/etc.

File: Library/Homebrew/test/testing_env.rb (hunk 4 above)
Comment (mikemcquaid) [inline Library/Homebrew/test/testing_env.rb]: It's not a hack, this is commonly done for coverage checking in Ruby projects.

Comment (xu-cheng) [inline Library/Homebrew/test/testing_env.rb]: By hack, I mean I rarely see other project to require all files using file glob. (On the hand, manually requiring all the code using `require "libname"` is quite common. )

My concerns are (a) this can raise hard to debug problem, as the order matters. (b) brew tests might give false postive result, because we don't require all codes like this outside test.

Comment (UniqMartin) [inline Library/Homebrew/test/testing_env.rb]: > Becuase this feels hack to me and likely to cause trouble: e.g. There may be conflicts, the order of files during require matters.

If this is the case, *this* feels like a hack to me. Correct me if I'm wrong, because my experience with Ruby is still limited, but we should fix those cases by renaming identifiers and fixing the `require`s at the top of those files instead of relying on order and only a subset being `require`d.

Is there an easy way to dump all the modules/classes/methods known to Ruby at a certain point during the program runtime? That would help to systematically check for some of those issues.

> If it doesn't missed out any coverage.

There are a lot of files we don't test yet. After this change, our coverage is down by about 10%. Not super nice, but at least the new number is much closer to reality.

Comment (mikemcquaid) [inline Library/Homebrew/test/testing_env.rb]: If it's only required for coverage tests: I think it's fine. Without something like this: our coverage % is completely wrong and we should remove Coveralls. Longer-term if we have every file included somewhere: :+1:

Comment (xu-cheng) [inline Library/Homebrew/test/testing_env.rb]: I would suggest don't include this even for coverage. Because, I think to get correct test result is more important than coverage. 

Also it's worthy nothing to mention, even something like this, the coverage percent is still wrong. We don't count coverage of codes used when running build or postinstall script. 

As for coverall, I think it's fine to keep it. Because I know the percent is wrong, at the same time, it does a good job to tracking which lines are already covered(notice to that it doesn't track what lines not covered)

**Comment by bfontaine** [comment]
I added some tests to compensate for the coverage % lost.

**Comment by xu-cheng** [comment]
@bfontaine I really think we shouldn't require all the files. Becuase it's pointless to have a coverage report with its test result being incorrect. 

Thankfully, after taking look at simplecov API, I found this. http://www.rubydoc.info/gems/simplecov/SimpleCov/Configuration#track_files-instance_method
I think this is right way to get accuracy coverage result without messing with the running code.

**Comment by UniqMartin** [comment]
> Thankfully, after taking look at simplecov API, I found this. http://www.rubydoc.info/gems/simplecov/SimpleCov/Configuration#track_files-instance_method
I think this is right way to get accuracy coverage result without messing with the running code.

@xu-cheng In theory, that's the perfect solution and by now I agree with you that we shouldn't simply `require` all those files. In practice, there are some problems with this suggestion:

- `track_files` requires SimpleCov 0.11.x, but we're using 0.10.0.
- We cannot update to SimpleCov 0.11.x as it has a bug with handling `root` that makes it completely useless to us (I tried, but it seems there's no easy workaround).
- `track_files` is not very well integrated and will partially override coverage information when merging multiple result sets (functionality we rely on for our integration tests).

**Comment by mikemcquaid** [comment]
> I really think we shouldn't require all the files. Becuase it's pointless to have a coverage report with its test result being incorrect.

The "test result being incorrect" seems to be speculative rather than actual so I'd rather see actual, current tests that fail/pass that shouldn't.

I'd really like to see this `require` method merged and then iterate if we can find a better method which still produces correct coverage.

File: Library/Homebrew/test/testing_env.rb [hunk 5]
```diff
@@ -20,6 +20,22 @@
...
+  # Ensure all Ruby files are `require`d for an accurate coverage measurement
+  library_base = File.expand_path("../..", __FILE__)
+  Dir["#{library_base}/**/*.rb"].each do |f|
+    f_rel = f[(library_base.length + 1)..-1] # path relative to `library_base`
+    # scripts
+    next if %w[build.rb postinstall.rb test.rb].include?(f_rel)
+    # mocked files
+    next if %w[config.rb global.rb].include?(f_rel)
+    # test files
+    next if f_rel.start_with?("test/")
```
Comment (UniqMartin) [inline Library/Homebrew/test/testing_env.rb]: I think it would be a good idea to extend this to `next if f_rel.start_with?("compat/", "test/", "vendor/")` to match the filters we have in `.simplecov` and because we don't care about the coverage of those files. The comment above would require adjustment for this, too.

Additionally, we should probably exclude `compat.rb` to avoid pulling them in indirectly. (And because I think we don't want this compatibility stuff to affect testing.)

**Comment by xu-cheng** [comment]
> The "test result being incorrect" seems to be speculative rather than actual so I'd rather see actual, current tests that fail/pass that shouldn't.

@mikemcquaid You can take a look at CI report for [this very PR](https://travis-ci.org/Homebrew/homebrew/jobs/103498334). What you will see is `brew tests` failed but `brew tests --coverage` passed. Noted the former one doesn't require all codes in such arbitrary way. It's also important to notice that many codes inside `compat` folder(and maybe other place) are order sensitive. 

> still produces correct coverage.

A note regardless this PR:  There are in fact many missing coverage when invoking `build.rb' and 'postinstall.rb` script.

File: Library/Homebrew/test/testing_env.rb (hunk 5 above)
Comment (bfontaine) [inline Library/Homebrew/test/testing_env.rb]: @UniqMartin shouldn’t we test `compat/` files as well?

**Comment by bfontaine** [comment]
That sounds hacky but we could maybe `require` all files in a separate job that doesn’t affect the current one (similarly to integration tests), so that we ensure all files are covered while at the same time running the tests in a clean environment (that is, without `require`-ing everything)?

**Comment by bfontaine** [comment]
@xu-cheng Good catch, this is indeed an issue that’s due to the `require`-all-the-things.

File: Library/Homebrew/test/testing_env.rb (hunk 5 above)
Comment (UniqMartin) [inline Library/Homebrew/test/testing_env.rb]: Maybe. But I guess the `add_filter "Homebrew/compat/"` and the `--no-compat` flag of `brew tests` is there for a reason?

**Comment by bfontaine** [comment]
To keep this PR simpler I removed the added tests and made a separate PR instead (#48281).

**Comment by UniqMartin** [comment]
I was able to fix a local copy of SimpleCov 0.11.1 to the point where its new `track_files` feature probably works as it should. Thus I was able to run `brew tests --no-compat --coverage` without `require`ing all the files, but still got coverage results for all 170 relevant Ruby files (the ones not used by any test get a coverage of 0%). The result is not very pretty, but it should be quite accurate: about 42% coverage.

**Comment by xu-cheng** [comment]
@UniqMartin I find a way to workaround the [root_filter issue](https://github.com/colszowka/simplecov/pull/437) with following code:
```ruby
SimpleCov.start do
  ...
  filters.clear
  add_filter { |src| !src.start_with?(SimpleCov.root) }
  ...
  track_files "#{SimpleCov.root}/**/*.rb"
end
```

But I haven't found a way to solve [merge issue](https://github.com/colszowka/simplecov/pull/436)

**Comment by UniqMartin** [comment]
> But I haven't found a way to solve [merge issue](https://github.com/colszowka/simplecov/pull/436)

Thanks for pointing me at this PR! I somehow failed to recognize it as what it is. My fix requires a relatively simple change to `SimpleCov::ArrayMergeHelper#merge_resultset`:

```diff
diff --git i/lib/simplecov/merge_helpers.rb w/lib/simplecov/merge_helpers.rb
index 339210d1..ca6b8a9f 100644
--- i/lib/simplecov/merge_helpers.rb
+++ w/lib/simplecov/merge_helpers.rb
@@ -2,6 +2,12 @@ module SimpleCov
   module ArrayMergeHelper
     # Merges an array of coverage results with self
     def merge_resultset(array)
+      # Don't merge with all-zero array to preserve nil values from actual runs.
+      if size == array.size
+        return array.dup if all? { |el| el == 0 }
+        return dup if array.all? { |el| el == 0 }
+      end
+
       new_array = dup
       array.each_with_index do |element, i|
         if element.nil? && new_array[i].nil?
```

(I should probably contribute that upstream, but I first wanted to get that working with our code. And this only happened about 15 minutes ago.)

**Comment by UniqMartin** [comment]
> That sounds hacky but we could maybe `require` all files in a separate job that doesn’t affect the current one (similarly to integration tests), so that we ensure all files are covered while at the same time running the tests in a clean environment (that is, without `require`-ing everything)?

If you can be bothered to try this, I think this is a good compromise between our desire to get coverage for all our files (excluding scripts for now) while eliminating the undesirable effects `require`-ing all of them for all other tests has. But I can sympathize if you don't want to waster your time on that.

Because eventually, we should be able to replace this by switching to a newer `simplecov` and using its `track_files` feature. But that might take a while, as there are issues that need to be resolved and a version that contains the fixes to be released (we might not want to run a custom-patched `simplecov`).

**Comment by mikemcquaid** [comment]
> Because eventually, we should be able to replace this by switching to a newer simplecov and using its track_files feature. But that might take a while, as there are issues that need to be resolved and a version that contains the fixes to be released (we might not want to run a custom-patched simplecov).

I'm :ok: with us running a patched one as long as the patch is submitted upstream (and ideally accepted).

**Comment by UniqMartin** [comment]
> I'm :ok: with us running a patched one as long as the patch is submitted upstream (and ideally accepted).

How would we go about doing that, i.e., what goes into `test/Gemfile` and where to put the patched `simplecov` files then? I'll try to submit my fixes to upstream or to contribute them to the already open PRs soon. I'm just having very limited time right now …

**Comment by mikemcquaid** [comment]
@UniqMartin You can point the `test/Gemfile` to (y)our own fork's branch. Don't worry about the timing, this isn't a huge rush or anything.

**Comment by bfontaine** [comment]
@UniqMartin I can handle that if you feel you don’t have the time.

**Comment by UniqMartin** [comment]
@bfontaine Feel free to do that, if you like. Here's a bit more information about the 0.11.0/1 issues:

The fix for the `track_files` feature with merging result sets is essentially in the diff in https://github.com/Homebrew/homebrew/pull/48250#issuecomment-173171990. The logic behind that fix is: If one of the files from a result set has all zeros, then it's probably generated by the `track_files` feature and we take the other one (which might or might not be from an actual run, but if it is, using it won't destroy the `nil` values for empty lines, comment lines, etc.). That should probably go into https://github.com/colszowka/simplecov/pull/436.

The other issue is with the `root_filter` being set before a custom `.simplecov` has the chance to alter `root` (which defaults to `pwd`). In our case this will cause all files outside of `Library/Homebrew/test/` to be filtered, always resulting in an empty coverage report. That's a regression caused by https://github.com/colszowka/simplecov/pull/396 and an attempt at a fix is in https://github.com/colszowka/simplecov/pull/437.

**Comment by bfontaine** [comment]
I believe I’ve now a working solution. I disabled `track_files` for integration tests to speed up the test suite (70 seconds on my mac vs. 3m51s) and had to make a few tweaks to have everything working with the patched `simplecov`.

We’re now down at ~46% coverage, which is quite bad but at least it’s now the correct number :+1:

File: Library/Homebrew/test/.simplecov [hunk 6]
```diff
@@ -3,14 +3,21 @@
 SimpleCov.start do
   tests_path = File.dirname(__FILE__)
 
-  minimum_coverage 60
+  minimum_coverage 40
   coverage_dir File.expand_path("#{tests_path}/coverage")
   root File.expand_path("#{tests_path}/../../")
 
   add_filter "Formula/"
+  add_filter "Taps/"
```
Comment (UniqMartin) [inline Library/Homebrew/test/.simplecov]: Any reason not to put this in alphabetical order?

**Comment by UniqMartin** [comment]
Thanks for all the work and for transforming the hints I left into a working solution! I really like the outcome! :smile_cat: :+1:

> I disabled `track_files` for integration tests to speed up the test suite (70 seconds on my mac vs. 3m51s) and […]

Cool! Simple change but massive impact. The slow-down caused by reading and writing megabytes of JSON over and over again is even more noticeable than I thought.

**Comment by bfontaine** [comment]
> The slow-down caused by reading and writing megabytes of JSON over and over again is even more noticeable than I thought.

Before I noticed it was also tracking files under `Taps/` the test suite needed half an hour to complete :disappointed:

File: Library/Homebrew/test/.simplecov (hunk 6 above)
Comment (bfontaine) [inline Library/Homebrew/test/.simplecov]: Nope.

**Comment by UniqMartin** [comment]
:+1: I'm really happy with the solution we arrived at, even if the path wasn't the shortest and most straightforward. The coverage drop is not great, but oh well … it can only get better. :smiley_cat:

**Comment by bfontaine** [comment]
@mikemcquaid @xu-cheng I’ll merge it if you’re happy with the current solution.

File: Library/Homebrew/test/Gemfile [hunk 7]
```diff
@@ -3,5 +3,15 @@ source "https://rubygems.org"
...
 gem "minitest", "~> 5.3"
 gem "rake", "~> 10.3"
-gem "simplecov", "~> 0.10.0", :require => false
-gem "coveralls", "0.8.9", :require => false
+# This is a patched version of the v0.11.1. Switch back to the stable version
+# when #436 and #437 have been merged upstream.
+# See https://github.com/Homebrew/homebrew/pull/48250#issuecomment-173171990
+gem "simplecov", "0.11.1",
+  :git => "https://github.com/Homebrew/simplecov.git",
+  :ref => "a010c8e92c2ea4e2c51f81b52c63b5f03926ccf8",
```
Comment (mikemcquaid) [inline Library/Homebrew/test/Gemfile]: Cab we point this to a branch instead and link directly to the PRs? Thanks!

**Comment by mikemcquaid** [comment]
Couple of final comments then: :+1:

File: Library/Homebrew/test/Gemfile (hunk 7 above)
Comment (bfontaine) [inline Library/Homebrew/test/Gemfile]: Done.

File: Library/Homebrew/test/Gemfile [hunk 8]
```diff
@@ -3,5 +3,17 @@ source "https://rubygems.org"
 gem "mocha", "~> 1.1"
 gem "minitest", "~> 5.3"
 gem "rake", "~> 10.3"
-gem "simplecov", "~> 0.10.0", :require => false
-gem "coveralls", "0.8.9", :require => false
+# This is a patched version of the v0.11.1. Switch back to the stable version
+# when #436 and #437 have been merged upstream.
+# https://github.com/colszowka/simplecov/pull/436
```
Comment (mikemcquaid) [inline Library/Homebrew/test/Gemfile]: Mind accepting your own PR for that one? It has conflicts and your patch is simple enough to be a separate PR.

Comment (bfontaine) [inline Library/Homebrew/test/Gemfile]: I assume you meant “submitting” :wink: 
Let’s wait until Monday for the 0.11.2 release, which will hopefully fix *both* issues.

Comment (mikemcquaid) [inline Library/Homebrew/test/Gemfile]: :cool:.

Comment (bfontaine) [inline Library/Homebrew/test/Gemfile]: 0.11.2 fixed https://github.com/colszowka/simplecov/pull/436 but not https://github.com/colszowka/simplecov/pull/437, for which we need to wait for 0.11.3. I updated our fork as well as this PR to use a patched 0.11.2 and I’m writing tests in the meantime to reduce the coverage drop.

**Comment by mikemcquaid** [comment]
Looks good :+1:

**Comment by UniqMartin** [comment]
:+1: Should we merge this and iterate on this later, once upstream resolves the remaining issue that currently affects us?

**Comment by mikemcquaid** [comment]
> Should we merge this and iterate on this later, once upstream resolves the remaining issue that currently affects us?

:+1: from me
